go-controller/pkg/libovsdbops/switch.go

```diff
@@ -201,33 +201,48 @@ func AddACLsToLogicalSwitchOps(nbClient libovsdbclient.Client, ops []libovsdb.Op
 // corresponding ops
 func RemoveACLsFromLogicalSwitchesWithPredicateOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation,
 	p switchPredicate, acls ...*nbdb.ACL) ([]libovsdb.Operation, error) {
-	sw := nbdb.LogicalSwitch{
-		ACLs: make([]string, 0, len(acls)),
-	}
+	aclUUIDs := make(map[string]bool, len(acls))
 	for _, acl := range acls {
-		sw.ACLs = append(sw.ACLs, acl.UUID)
+		aclUUIDs[acl.UUID] = true
 	}
-	opModel := operationModel{
-		Model:            &sw,
-		ModelPredicate:   p,
-		OnModelMutations: []interface{}{&sw.ACLs},
-		ErrNotFound:      false,
-		BulkOp:           true,
+	switches, err := FindLogicalSwitchesWithPredicate(nbClient, p)
+	if err != nil {
+		return ops, fmt.Errorf("failed to remove acls from the logical switches with predicate, "+
+			"predicate search failed: %w", err)
+	}
+	opModels := make([]operationModel, 0, len(switches)+1)
+	for _, sw := range switches {
+		sw := sw
+		aclsToRemove := []string{}
+		for _, swACLUUID := range sw.ACLs {
+			if aclUUIDs[swACLUUID] {
+				aclsToRemove = append(aclsToRemove, swACLUUID)
+			}
+		}
+		sw.ACLs = aclsToRemove
+		opModel := operationModel{
+			Model:            sw,
+			OnModelMutations: []interface{}{&sw.ACLs},
+			ErrNotFound:      false,
+			BulkOp:           true,
+		}
+		opModels = append(opModels, opModel)
 	}
-
 	m := newModelClient(nbClient)
-	return m.DeleteOps(ops, opModel)
+	return m.DeleteOps(ops, opModels...)
```

Comment:
the intention is to speed up egress firewall sync, we will call `RemoveACLsFromLogicalSwitchesWithPredicate` for all egress firewall acls on every restart, and it will have a real effect only once, and will try to delete acl that don't exist on switched for the other calls.
I am not sure how it is more efficient to perform this on the server side, if just decoding a long list+waiting for the transaction reply will probably take more time than this check?

Reply:
You are thinking that the current use of `RemoveACLsFromLogicalSwitchesWithPredicate` is the only one there will ever be.

Would it be another option is to use the predicate instead? So look in the predicate if any of the passed ACLs is in the switch? The transaction might include ACLs that are not on the switch, but that really does not apply to your concern above. Once they have been removed, this will result in no transaction as well.

Reply:
I am not sure I see why exactly it is a better option: the result will be the same, since predicate will be evaluated before the ops are returned, and in case 1 acl of 1M acls is present on the switch, we will send all 1M acls in the ops.
there must be some logical benefit, can you please clarify?
I see the potential problem in doing this search in general, because if the acl is added to the switch in the same transaction, then it won't be deleted by these ops with the optimization, but it is a different problem

Reply:
In general, searching for UUIDs in a column is more efficiently done in the server because they keep the UUIDs ordered, do a binary search and maybe other optimizations.

So it is better for us to not do that search in general, which is what is happening in this change. 

If done in a predicate, it's no happening in general. Just when that predicate is specifically used. So we can still offer other uses of `RemoveACLsFromLogicalSwitchesWithPredicate` that don't to do that and benefit from a better operation in the server.

Reply:
just to search UUIDs on the server side, we need to include overhead of encoding the request, sending it to the server (afaik client is single-threaded, so the transaction will wait in the queue until the previous transactions are executed?), here happens a fast server-side evaluation, and a reply. 

so for 1switch 1M acls transaction:
worst-case with client-side evaluation, all acls are present on the switch, we spend ~0.35s extra time to compare acls
worst-case without client-side evaluation, no acls are present on the switch, we spend ~4 seconds to transact empty ops

Also based on your analysis encoding the request takes the most time, so when we just remove some acls from the ops, we save a lot of time. e.g. if we only leave 500`000 ACLs from 1M to transact (thinking they are present on the switch), the same transaction takes ~2.2s instead of 4s for 1M acls.
maybe if (not sure if that is possible) we get to a point where sending a big request to the server is actually more efficient than optimizing it on the client side, we can rework that part? maybe I am missing something here, please let me know what you think

Reply:
> worst-case without client-side evaluation, no acls are present on the switch, we spend ~4 seconds to transact empty ops

Not with the predicate I described above. If none of those ACLs are present on the switch, the predicate would evaluate false and no transaction would be sent.

> Also based on your analysis encoding the request takes the most time, so when we just remove some acls from the ops, we save a lot of time. 

True but I am trying to reach a compromise where we keep `RemoveACLsFromLogicalSwitchesWithPredicate` in it's most reasonable form considering the migration as an exceptional case and that **other, future uses** of `RemoveACLsFromLogicalSwitchesWithPredicate` might be more reasonable. We are basically changing `RemoveACLsFromLogicalSwitchesWithPredicate` for a exceptional transaction that will only ever happen once on some deployments.

So I don't know how important this consideration you are making is. The first time the migration runs it might be a heavy operation anyway that we keep in check with batching. Yes, here we might save some time if some ACLs are in some switches but not in others, do you have an estimation how different migrated ACLs are distributed? After a first migration, then nothing matters because there is nothing left to migrate.

And yet another different option might be a method `RemoveACLsFromAllLogicalSwitches` that would internally use `FindLogicalSwitchesWithPredicate` with the predicate I described above and then uses an op as described in the first comment. This would work because while all ACLs might not be present in all switches, certainly every ACL will be present in at least one switch.